libraries/AP_Baro/AP_Baro_UAVCAN.cpp

```diff
@@ -5,113 +5,181 @@
+        AP_HAL::panic("UAVCAN Baro subscriber start problem\n\r");
+        return;
+    }
 }
 
-AP_Baro_UAVCAN::~AP_Baro_UAVCAN()
+void AP_Baro_UAVCAN::handle_pressure(AP_UAVCAN* ap_uavcan, uint8_t node_id, const PressureCb &cb)
 {
-    if (!_initialized) {
+    AP_Baro_UAVCAN* driver = get_uavcan_backend(ap_uavcan, node_id, true);
+    if (driver == nullptr) {
         return;
     }
-    
-    AP_UAVCAN *ap_uavcan = AP_UAVCAN::get_uavcan(_manager);
-    if (ap_uavcan == nullptr) {
-        return;
+    if (driver->_sem_baro->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {        
+        driver->_pressure = cb.msg->static_pressure;
+        driver->new_pressure = true;
+        driver->_sem_baro->give();
     }
-    
-    ap_uavcan->remove_baro_listener(this);
-    delete _sem_baro;
-    
-    debug_baro_uavcan(2, "AP_Baro_UAVCAN destructed\n\r");
 }
 
-AP_Baro_Backend *AP_Baro_UAVCAN::probe(AP_Baro &baro)
+void AP_Baro_UAVCAN::handle_temperature(AP_UAVCAN* ap_uavcan, uint8_t node_id, const TemperatureCb &cb)
 {
+    AP_Baro_UAVCAN* driver = get_uavcan_backend(ap_uavcan, node_id, false);
+    if (driver == nullptr) {
+        return;
+    }
+    if (driver->_sem_baro->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+        driver->_temperature = cb.msg->static_temperature;
+        driver->_sem_baro->give();
+    }
+}
+
+
 
-    if (AP_BoardConfig_CAN::get_can_num_ifaces() == 0) {
+AP_Baro_UAVCAN* AP_Baro_UAVCAN::get_uavcan_backend(AP_UAVCAN* ap_uavcan, uint8_t node_id, bool create_new)
+{
+    AP_Baro_UAVCAN* driver = nullptr;
+    if (ap_uavcan == nullptr) {
         return nullptr;
     }
-
-    AP_Baro_UAVCAN *sensor;
-    for (uint8_t i = 0; i < MAX_NUMBER_OF_CAN_DRIVERS; i++) {
-        AP_UAVCAN *ap_uavcan = AP_UAVCAN::get_uavcan(i);
-        if (ap_uavcan == nullptr) {
-            continue;
-        }
-        
-        uint8_t freebaro = ap_uavcan->find_smallest_free_baro_node();
-        if (freebaro == UINT8_MAX) {
-            continue;
+    for (uint8_t i = 0; i < BARO_MAX_DRIVERS; i++) {
+        if (_detected_modules[i].driver != nullptr &&
+            _detected_modules[i].ap_uavcan == ap_uavcan && 
+            _detected_modules[i].node_id == node_id) {
+            return _detected_modules[i].driver;
         }
-        sensor = new AP_Baro_UAVCAN(baro);
-        if (sensor->register_uavcan_baro(i, freebaro)) {
-            debug_baro_uavcan(2, "AP_Baro_UAVCAN probed, drv: %d, node: %d\n\r", i, freebaro);
-            return sensor;
-        } else {
-            delete sensor;
+    }
+    
+    if (driver == nullptr && create_new) {
+        if (take_registry()) {
+            bool already_detected = false;
+            //Check if there's an empty spot for possible registeration
+            for (uint8_t i = 0; i < BARO_MAX_DRIVERS; i++) {
+                if (_detected_modules[i].ap_uavcan == ap_uavcan && _detected_modules[i].node_id == node_id) {
+                    //Already Detected
+                    already_detected = true;
```

Comment:
@OXINARF It is like this because the removal and addition might not happen in order, this will be the case for dynamic allocation and unallocation.

Reply:
@bugobliterator I don't understand what you are saying. Why can't you just break? The variable isn't going to change value again by continuing the loop.
Also, the lowest empty index can't change because you are holding the registry semaphore the entire time.